### PR TITLE
Added Dinitz's and Push-Relabel algorithm

### PR DESCRIPTION
**Pull request by pranavtbhat**
I've added two new flow algorithms, Dinitz's algorithm and a FIFO version of the Push-Relabel algorithm. `dinitz_impl` and `push_relabel` should be faster than the existing `edmund_karp_impl`, but seem to be slightly slower. I guess I haven't been able to optimize them. 

```julia
using LightGraphs
using Base.Test

flow_graph = readgraph("/home/pranav/.julia/v0.4/LightGraphs/test/testdata/graph-5k-50k.jgz")["graph-5000-50000"]

capacity_matrix = zeros(5000,5000)

for u in vertices(flow_graph)
    for v in flow_graph.fadjlist[u]
        capacity_matrix[u,v] = rand(1:1000)
    end
end


gc()
@time f1,F1 = maximum_flow(flow_graph,1,5000,capacity_matrix,algorithm=EdmundKarpAlgorithm())
gc()
@time f2,F2 = maximum_flow(flow_graph,1,5000,capacity_matrix,algorithm=DinitzAlgorithm())
gc()
@time f3,F3 = maximum_flow(flow_graph, 1, 5000,capacity_matrix,algorithm=PushRelabelAlgorithm())
```
This gave:
```
  0.052088 seconds (53.11 k allocations: 208.144 MB, 3.81% gc time)
  0.061093 seconds (53.91 k allocations: 201.796 MB, 3.40% gc time)
  0.073750 seconds (52.73 k allocations: 201.557 MB, 2.70% gc time)
``` 

Oh and how do I write tests for void-methods (methods returning `nothing`) so that coverage includes it?

### CODE REVIEW

**Comment by sbromberger** [comment]
So, a couple of things:

1) don't use references to the internal datastructures. For `g.fadjlist[x]`, you can use `fadj(g,x)`.
2) `test foo(...) == nothing` should work.

File: src/flow/maximum_flow.jl [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    source::Int,                           # the source vertex
+    target::Int,                           # the target vertex
+    capacity_matrix::AbstractArray{T,2}=   # edge flow capacities
+        DefaultDistance();
+    algorithm::AbstractFlowAlgorithm =     # keyword argument for algorithm
+        EdmundKarpAlgorithm()
+    )
+    residual_graph, capacity_matrix = residual(flow_graph, capacity_matrix)
+
+    if typeof(algorithm) == EdmundKarpAlgorithm
```
Comment (jpfairbanks) [inline src/flow/maximum_flow.jl]: Instead of this pythonic type switch can you use mutliple dispatch?

Comment (sbromberger) [inline src/flow/maximum_flow.jl]: that is, you *should* use MD. See how the traversals work (they do this with BFS and DFS).

**Comment by jpfairbanks** [comment]
As for your timings, the elapsed times are too small to rely on one sample, you will need to run either on bigger graphs or take more samples (preferably both). Also you should track down which function is doing so much allocating 53K allocations but only 200MB seems weird based on the code I skimmed.

**Comment by CarloLucibello** [comment]
Also, while  they have the same complexity for sparse graphs, you could try to use them on a complete graph. The different scalings with n should be immediately noticeable

**Comment by jpfairbanks** [comment]
Do you need weights in order to get the right behavior on the complete graph? I could see the symmetry causing run time to be much better than the guarantee on such a special graph.

**Comment by pranavtbhat** [comment]
Sorry, I'll avoid using internal data structures in the future. I'll add some tests for the void-functions in `push_relabel` soon. 

I will try running the algorithm on denser graphs and also use the @profile macro to see what's doing the allocations.

**Comment by sbromberger** [comment]
Also will need more test coverage (goal is 100%!)

**Comment by pranavtbhat** [comment]
I added tests for all the methods in `push_relabel` , switched over to multiple dispatch. The build seems to be failing on 0.5. 
```julia
ERROR: LoadError: LoadError: invalid doc expression:
@doc "..." "maximum_flow\nMethod for Edmund Karp's Algorithm\n"
```

I tested the performance of all three algorithms fo large datasets (5000 vertices).
image
(The x-axis is density and not number of edges)

Varying the number of vertices between (1000 and 8000) showed a similar plot.

File: src/flow/dinitz.jl [hunk 2]
```diff
@@ -0,0 +1,141 @@
+"""dinitz_impl
+Computes the maximum flow between the source and target vertexes in a flow
```
Comment (sbromberger) [inline src/flow/dinitz.jl]: You don't need the function name in the docstring. Just start with "Computes..."

File: src/flow/edmund_karp.jl [hunk 3]
```diff
@@ -1,18 +1,24 @@
-"""Computes the maximum flow between the source and target vertexes in a flow
+"""edmund_karp_impl
+Computes the maximum flow between the source and target vertexes in a flow
 graph using [Edmond Karp's](https://en.wikipedia.org/wiki/Edmonds%E2%80%93Karp_algorithm)
```
Comment (sbromberger) [inline src/flow/edmund_karp.jl]: Edmond? Why is the function called `edmund`?

File: src/flow/maximum_flow.jl [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            add_edge!(residual_graph, v, u)
+        end
+    end
+
+    return residual_graph
+end
+
+"""
+Generic maximum_flow function that can use one of the three flow algorithms:
+1. Endmond Karp's algorithm: O(VE^2).
```
Comment (sbromberger) [inline src/flow/maximum_flow.jl]: Check the name again :)

File: src/flow/edmund_karp.jl (hunk 3 above)
Comment (pranavtbhat) [inline src/flow/edmund_karp.jl]: Wow, it's supposed to be Edmonds-Karp. Need to make a lot of changes now :(

File: src/flow/dinitz.jl (hunk 2 above)
Comment (pranavtbhat) [inline src/flow/dinitz.jl]: Removed the function name from all docstrings.

File: src/flow/edmund_karp.jl (hunk 3 above)
Comment (pranavtbhat) [inline src/flow/edmund_karp.jl]: I changed Dinitz to Dinic as well. Escaped all ' s

**Comment by sbromberger** [comment]
Awesome - looks like Travis is happy. Can we get to 100% coverage?

**Comment by pranavtbhat** [comment]
Fixed the coverage-patch check.
.

File: src/flow/edmonds_karp.jl [hunk 5]
```diff
@@ -1,18 +1,24 @@
-"""Computes the maximum flow between the source and target vertexes in a flow
-graph using [Edmond Karp's](https://en.wikipedia.org/wiki/Edmonds%E2%80%93Karp_algorithm)
+"""
+Computes the maximum flow between the source and target vertexes in a flow
+graph using [Edmonds Karp\'s](https://en.wikipedia.org/wiki/Edmondss%E2%80%93Karp_algorithm)
```
Comment (sbromberger) [inline src/flow/edmonds_karp.jl]: This is `Edmonds-Karp` (two people; one named Jack Edmonds, the other Richard Karp).

File: src/flow/maximum_flow.jl [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+type DefaultCapacity <: AbstractArray{Int, 2}
+    flow_graph::LightGraphs.DiGraph
+    nv::Int
+    DefaultCapacity(flow_graph::LightGraphs.DiGraph) = new(flow_graph, nv(flow_graph))
+end
+
+getindex(d::DefaultCapacity, s::Int, t::Int) = if has_edge(d.flow_graph, s , t) 1 else 0 end
+size(d::DefaultCapacity) = (d.nv, d.nv)
+transpose(d::DefaultDistance) = d
```
Comment (sbromberger) [inline src/flow/maximum_flow.jl]: Should this be DefaultCapacity instead of DefaultDistance?

Comment (pranavtbhat) [inline src/flow/maximum_flow.jl]: Oops. I'll change that. But considering that DefaultCapacity isn't symmetric, the transpose may not be the same matrix. Is there any function to reverse the graph (reverse all the edges?)

Comment (sbromberger) [inline src/flow/maximum_flow.jl]: Yes. There's a `reverse(g::DiGraph)` and an in-place `reverse!(g::DiGraph)`.

**Comment by sbromberger** [comment]
@pranavtbhat - Do you need to make a change based on https://github.com/JuliaGraphs/LightGraphs.jl/pull/201#discussion_r43778021, or are we good to merge?

**Comment by pranavtbhat** [comment]
I think I'll change the transpose and ctranspose implementations. Oh and do you want me to run a rebase to squash commits?

**Comment by pranavtbhat** [comment]
I think we are good to merge unless a rebase is required.

**Comment by sbromberger** [comment]
no rebase required. :) Thanks very much for this.

**Pull request merged**